### PR TITLE
src,qa: Upgrade to mypy 0.901

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Fixed:

* rook/rook_cluster.py:591: error: List comprehension has incompatible type List[Union[str, Any]]; expected List[Optional[Device]]
* dashboard/plugins/feature_toggles.py:104: error: List comprehension has incompatible type List[str]; expected List[Features]

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by tchaikov — COMMENTED** [review]
LGTM, once
```
rook/rook_cluster.py: note: In function "add_osds":
rook/rook_cluster.py:591: error: List comprehension has incompatible type List[Union[str, Any]]; expected List[Optional[Device]]
Found 1 error in 1 file (checked 23 source files)
```
is addressed

**Review by tchaikov — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
this needs a review from @ceph/dashboard

**Comment by sebastian-philipp** [comment]
this needs a review from @ceph/dashboard

File: src/pybind/mgr/dashboard/plugins/feature_toggles.py
Comment (epuertat) [inline src/pybind/mgr/dashboard/plugins/feature_toggles.py:104]: Mmm I don't think this is equivalent:

```py
[<Features.RBD: 'rbd'>, <Features.MIRRORING: 'mirroring', ...] != ['rbd', 'mirroring', ...]
```

**Review by epuertat — CHANGES_REQUESTED** [review]
There's an issue with the enum there

File: src/pybind/mgr/dashboard/plugins/feature_toggles.py
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/plugins/feature_toggles.py:104]: 1. `def cmd` takes the features as arguments: `features: Optional[List[Features]] = None`. Which means `features` in this line is of type `List[Features]`. 

2. In order for `features or [f.value for f in Features]` to make sense, we should make sure that the first part of `or` has the same type as the second part. `[f.value for f in Features]` is of type `List[str]`. So right now, the `feature` is either of type `Feature` or of type `str` based on the value of `features`. 

The current code only works, because `"{}".format(feature)` returns the same value for both `Feature` and `str`.

**Comment by sebastian-philipp** [comment]
friendly ping @epuertat

File: src/pybind/mgr/dashboard/plugins/feature_toggles.py
Comment (epuertat) [reply src/pybind/mgr/dashboard/plugins/feature_toggles.py:104]: I think part of the confusion here is that the `cmd(.., features: Optional[List[Features]], ...)` type hint is not correct:
* There's no way for a `MonCommand` to cast a CLI string to a `Feature` Enum: it'll always be a `str` (or a list of `str`s): "rbd,iscsi,rbdmirror".
* [The proper syntax](https://www.python.org/dev/peps/pep-0586/#interactions-with-enums-and-exhaustiveness-checks) here would have been `features: Optional[List[Literal[Features]]]`.

**Comment by epuertat** [comment]
> friendly ping @epuertat

Thanks, @sebastian-philipp ! I didn't see this one.

**Comment by sebastian-philipp** [comment]
rados/cephadm passed https://pulpito.ceph.com/swagner-2021-03-09_14:01:08-rados:cephadm-wip-swagner3-testing-2021-03-09-1022-distro-basic-smithi/

File: src/pybind/mgr/dashboard/plugins/feature_toggles.py
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/plugins/feature_toggles.py:104]: Sure? @tchaikov added support to annotate CLI commands with Enums and then generate the ceph command description out of the python types. 

See 

https://github.com/ceph/ceph/blob/f744878d82d96c3fd9d11ef5bd74abf321a62b72/src/pybind/mgr/orchestrator/module.py#L35-L39

and 

https://github.com/ceph/ceph/blob/f744878d82d96c3fd9d11ef5bd74abf321a62b72/src/pybind/mgr/orchestrator/module.py#L347-L348

Comment (epuertat) [reply src/pybind/mgr/dashboard/plugins/feature_toggles.py:104]: Something is broken here.

Let's assume:
```py
class Features(Enum):
    RBD = 'rbd'
    RGW = 'rgw'
    ...
```

If `feature` is the `Feature.<enum>.name` item ('RBD', 'RGW', ...), then the following line is broken (as it should be 'rbd', 'rgw'):
```py
msg += ["Feature '{}': {}".format(
   feature
```

If `feature` is the `Feature.<enum>.value` item ('rbd', 'rgw', ...), then the following line is broken (as it should be 'RBD', 'RGW'):
```py
    enabled = mgr.get_module_option(self.OPTION_FMT.format(feature))
```

File: src/cephadm/tox.ini
Comment (tchaikov) [inline src/cephadm/tox.ini:44]: i assume that we stick with 0.812 instead of 0.901 for good reasons.

**Review by tchaikov — APPROVED** [review]

File: src/cephadm/tox.ini
Comment (sebastian-philipp) [reply src/cephadm/tox.ini:44]: no. overlooked it

**Comment by sebastian-philipp** [comment]
```
ERROR: /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/iscsi.py Imports are incorrectly sorted and/or formatted.
ERROR: /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/rbd_mirroring.py Imports are incorrectly sorted and/or formatted.
ERROR: /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/plugins/debug.py Imports are incorrectly sorted and/or formatted.
ERROR: /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/plugins/feature_toggles.py Imports are incorrectly sorted and/or formatted.
ERROR: /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/plugins/plugin.py Imports are incorrectly sorted and/or formatted.
Skipped 3 files
ERROR: InvocationError for command /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/.tox/lint/bin/isort . --check (exited with code 1)
```

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
@epuertat hi Ernesto, i think your comment https://github.com/ceph/ceph/pull/39624#discussion_r584861887 has been resolved, could you please take another look?

File: qa/tox.ini
Comment (epuertat) [inline qa/tox.ini:21]: I wonder if it'd make sense to have a separate `requirements-typing.txt` for each sub-project, or even a common one for all (in order to ensure as many type definitions as possible are always installed and used).

**Review by epuertat — APPROVED** [review]
LGTM! Thanks @tchaikov

**Comment by epuertat** [comment]
> @epuertat hi Ernesto, i think your comment [#39624 (comment)](https://github.com/ceph/ceph/pull/39624#discussion_r584861887) has been resolved, could you please take another look?

@tchaikov yup, that was fixed a while ago!

File: qa/tox.ini
Comment (sebastian-philipp) [reply qa/tox.ini:21]: #41859

**Pull request merged**
